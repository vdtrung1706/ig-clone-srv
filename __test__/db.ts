import dotenv from 'dotenv';
import mongoose from 'mongoose';

dotenv.config();

export async function connectToDb(endoint: string): Promise<void> {
  await mongoose.connect(process.env.DB_TESTING_URL + `-${endoint}`);
}

export async function dropDb(): Promise<void> {
  if (mongoose.connection.readyState === 1) {
    await mongoose.connection.db
      .dropDatabase()
      .catch((err) => console.log(err));
  }
}

export async function disconnectDb(): Promise<void> {
  await mongoose.disconnect();
}

export function generateMongooseId(): mongoose.Types.ObjectId {
  return new mongoose.Types.ObjectId();
}
