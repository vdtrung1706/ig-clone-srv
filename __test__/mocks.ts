export const userMock = {
  email: 'vdtrung@gmail.com',
  username: 'vdtrung',
  fullName: 'Vu Dinh Trung',
  password: 'trung1234',
};

export const postMock = {
  image: 'http://mock_img.png',
  caption: 'post number 1',
  hashtags: ['#one', '#two'],
};
