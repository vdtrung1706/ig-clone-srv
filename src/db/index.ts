import mongoose from 'mongoose';
import Comment from '../models/comment';
import Post from '../models/post';
import User from '../models/user';
import Follow from '../models/follow';

export function connect(url: string): Promise<typeof mongoose> {
  return mongoose.connect(url);
}

export const models = {
  User,
  Post,
  Comment,
  Follow,
};
