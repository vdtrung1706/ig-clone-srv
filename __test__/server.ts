/* eslint-disable @typescript-eslint/ban-types */
import { convertNodeHttpToRequest, runHttpQuery } from 'apollo-server-core';
import { ApolloServer, ExpressContext } from 'apollo-server-express';
import express from 'express';
import { DocumentNode, ExecutionResult, print } from 'graphql';
import httpMocks, { RequestOptions, ResponseOptions } from 'node-mocks-http';
import { resolvers, typeDefs } from '../src/api';
import { createToken, getUserFromToken } from '../src/api/middlewares/auth';
import * as db from '../src/db';
import { IContext } from '../src/interfaces/IContext';
import { IUser } from '../src/interfaces/IUser';

// Create a test client from a lib: apollo-server-integration-testing

export type StringOrAst = string | DocumentNode;

export type Options<T extends object> = { variables?: T };

export type TestClientConfig = {
  apolloServer: ApolloServer;
  extendMockRequest?: RequestOptions;
  extendMockResponse?: ResponseOptions;
};

export type TestQuery = <T extends object = {}, V extends object = {}>(
  operation: StringOrAst,
  options?: Options<V>
) => Promise<ExecutionResult<T>>;

export type TestSetOptions = (options: {
  request?: RequestOptions;
  response?: ResponseOptions;
}) => void;

const mockRequest = (options: RequestOptions = {}) =>
  httpMocks.createRequest({
    method: 'POST',
    ...options,
  });

const mockResponse = (options: ResponseOptions = {}) =>
  httpMocks.createResponse(options);

export function createTestClient({
  apolloServer,
  extendMockRequest,
  extendMockResponse,
}: TestClientConfig): {
  query: TestQuery;
  mutate: TestQuery;
  setOptions: TestSetOptions;
} {
  const app = express();
  apolloServer.start().then(() => apolloServer.applyMiddleware({ app }));

  let mockRequestOptions = extendMockRequest;
  let mockResponseOptions = extendMockResponse;

  /**
   * Set the options after TestClient creation
   * Useful when you don't want to create a new instance just for a specific change in the request or response.
   *  */
  const setOptions: TestSetOptions = ({
    request,
    response,
  }: {
    request?: RequestOptions;
    response?: ResponseOptions;
  }) => {
    if (request) {
      mockRequestOptions = request;
    }
    if (response) {
      mockResponseOptions = response;
    }
  };

  const test: TestQuery = async <T extends object = {}, V extends object = {}>(
    operation: StringOrAst,
    { variables }: Options<V> = {}
  ) => {
    const req = mockRequest(mockRequestOptions);
    const res = mockResponse(mockResponseOptions);

    const graphQLOptions = await apolloServer.createGraphQLServerOptions(
      req,
      res
    );

    const { graphqlResponse } = await runHttpQuery([req, res], {
      method: 'POST',
      options: graphQLOptions,
      query: {
        query: typeof operation === 'string' ? operation : print(operation),
        variables,
      },
      request: convertNodeHttpToRequest(req),
    });

    return JSON.parse(graphqlResponse) as T;
  };

  return {
    query: test,
    mutate: test,
    setOptions,
  };
}

export function createServer(userMock: IUser): ApolloServer<ExpressContext> {
  const server = new ApolloServer({
    typeDefs,
    resolvers,
    context: async ({ req, res }): Promise<IContext> => {
      const models = db.models;
      if (userMock) {
        return {
          res,
          req,
          user: userMock,
          models,
          createToken,
        };
      }

      const token = req.headers.authorization;
      const user = await getUserFromToken(token);

      return {
        res,
        req,
        user,
        models,
        createToken,
      };
    },
  });

  return server;
}

export const createTestServer = (
  userMock?: IUser
): {
  query: TestQuery;
  mutate: TestQuery;
  setOptions: TestSetOptions;
} =>
  createTestClient({
    apolloServer: createServer(userMock),
  });
