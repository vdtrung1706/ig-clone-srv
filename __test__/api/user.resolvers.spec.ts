import Post from '../../src/models/post';
import User from '../../src/models/user';
import * as db from '../db';
import { postMock, userMock } from '../mocks';
import { createTestServer } from '../server';
import { gql } from 'apollo-server-core';

describe('User Resolvers', () => {
  beforeAll(() => db.connectToDb('user'));

  afterAll(db.dropDb);
  afterAll(db.disconnectDb);

  it('query user by id & bookmarks should be in a corrected form', async () => {
    const post = await Post.create(postMock);
    const user = await User.create({
      ...userMock,
      bookmarks: [{ _id: post.id }],
    });

    const server = createTestServer(user);

    const GET_USER = gql`
      query userQuery($id: String!) {
        user(id: $id) {
          email
          username
          fullName
          bookmarks {
            caption
          }
        }
      }
    `;

    const { data } = await server.query(GET_USER, {
      variables: {
        id: user.id,
      },
    });

    expect(data).toMatchSnapshot();
  });
});
