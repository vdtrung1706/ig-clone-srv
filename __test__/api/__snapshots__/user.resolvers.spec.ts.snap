// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Resolvers query user by id & bookmarks should be in a corrected form 1`] = `
Object {
  "user": Object {
    "bookmarks": Array [
      Object {
        "caption": "post number 1",
      },
    ],
    "email": "vdtrung@gmail.com",
    "fullName": "Vu Dinh Trung",
    "username": "vdtrung",
  },
}
`;
